test/e2e/watchcache/watchcache_enabled_test.go

```diff
@@ -202,3 +214,70 @@ func collectCacheHitsFor(ctx context.Context, t *testing.T, config *rest.Config,
 	}
 	return totalCacheHits, prefixCacheHit
 }
+
+const resyncPeriod = 10 * time.Hour
+const byWorkspace = "byWorkspace"
+
+func testDynamicDiscoverySharedInformerFactory(ctx context.Context, t *testing.T, restCfg *rest.Config, expectedGVR schema.GroupVersionResource, expectedResName string, expectedClusterName logicalcluster.Name) {
+	// since wildcard request are only allowed against a shard
+	// create a cfg that points to the root shard and use it to create ddsif
+	kcpClusterClient, err := kcpclientset.NewClusterForConfig(restCfg)
+	require.NoError(t, err)
+	cfg := framework.ShardConfig(t, kcpClusterClient, "root", restCfg)
+	kcpClusterClient, err = kcpclientset.NewClusterForConfig(cfg)
+	require.NoError(t, err)
+	kcpClient := kcpClusterClient.Cluster(logicalcluster.Wildcard)
+	kubeClusterClient, err := kubernetesclientset.NewClusterForConfig(cfg)
+	require.NoError(t, err)
+	metadataClusterClient, err := metadataclient.NewDynamicMetadataClusterClientForConfig(cfg)
+	require.NoError(t, err)
+	kcpSharedInformerFactory := kcpexternalversions.NewSharedInformerFactoryWithOptions(kcpClient, resyncPeriod)
+	ddsif := informer.NewDynamicDiscoverySharedInformerFactory(
+		kcpSharedInformerFactory.Tenancy().V1alpha1().ClusterWorkspaces().Lister(),
+		kubeClusterClient.DiscoveryClient,
+		metadataClusterClient.Cluster(logicalcluster.Wildcard),
+		func(obj interface{}) bool { return true }, 10*time.Second,
+	)
+	err = ddsif.AddIndexers(cache.Indexers{byWorkspace: indexByWorkspace})
+	require.NoError(t, err)
+
+	t.Log("Starting KCP Shared Informer Factory")
+	kcpSharedInformerFactory.Start(ctx.Done())
+	t.Log("Waiting for KCP Shared Informer Factory to sync caches")
+	kcpSharedInformerFactory.WaitForCacheSync(ctx.Done())
+	t.Log("Starting DynamicDiscoverySharedInformerFactory")
+	ddsif.Start(context.Background())
+
+	t.Logf("Checking if DynamicDiscoverySharedInformerFactory has %v with name %v in cluster %v", expectedGVR.String(), expectedResName, expectedClusterName)
+	framework.Eventually(t, func() (success bool, reason string) {
+		listers, notSynced := ddsif.Listers()
+		for _, ns := range notSynced {
+			klog.Infof("DynamicDiscoverySharedInformerFactory hasn't synced %v yet", ns.String())
+		}
+
+		for gvr, lister := range listers {
+			obj, err := lister.List(labels.Everything())
+			require.NoError(t, err)
+			if gvr.String() != expectedGVR.String() {
+				continue
+			}
+			for _, o := range obj {
+				u := o.(*unstructured.Unstructured)
+				if u.GetName() == expectedResName && u.GetClusterName() == expectedClusterName.String() {
+					return true, ""
+				}
+			}
+		}
+		return false, fmt.Sprintf("haven't found %v with name %v in %v cluster", expectedGVR, expectedResName, expectedClusterName)
+	}, wait.ForeverTestTimeout*2, time.Millisecond*100, "DynamicDiscoverySharedInformerFactory hasn't observed %v with name %v in %v cluster", expectedGVR, expectedResName, expectedClusterName)
```

Comment:
Why forever * 2?

Reply:
I think that `wait.ForeverTestTimeout` is not enough. I could increase the QPS and lower the poll timeout to speed up the test.

Reply:
done.